### PR TITLE
AspNet Core Settings - Basic implementation

### PR DESCRIPTION
**Pull request by StefanKert**
The title is pretty self-explaining. Please take a look at the current implementation and add things if you find something that makes no sense, or tell me to change it :). 

There are still some points that need to be discussed:

- Should we register a Method that refreshes the Settings on each file change?
- Should we add a validation for the settings file? Especially if someone adds an errorstore that is not supported it would be better to fail early. Maybe there are further settings that need to be validated.
- I haven´t found a similar event for a post deserialize for the configuration implementation in asp.net core but maybe you have an idea? Currently, I am calling all initialize methods after the Binding happened.

I really tried to add up to the current ConfigSettings class in StackExchange.Exceptional so maybe we can share some code here, but this would need some workarounds, so I think it is a good solution to use a separate class structure for this.

Thanks for your feedback. Really appreciate any input :)

### CODE REVIEW

File: src/StackExchange.Exceptional.AspNetCore/StackExchange.Exceptional.AspNetCore.csproj [hunk 1]
```diff
@@ -11,6 +11,21 @@
...
+    <Compile Update="ConfigSettings.ErrorStore.cs">
+      <DependentUpon>ConfigSettings.cs</DependentUpon>
+    </Compile>
+    <Compile Update="ConfigSettings.IgnoreErrors.cs">
+      <DependentUpon>ConfigSettings.cs</DependentUpon>
+    </Compile>
+    <Compile Update="ConfigSettings.LogFilters.cs">
+      <DependentUpon>ConfigSettings.cs</DependentUpon>
+    </Compile>
+  </ItemGroup>
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/StackExchange.Exceptional.AspNetCore.csproj]: Use a glob here:
```xml
<ItemGroup>
  <Compile Update="ConfigSettings.*.cs">
    <DependentUpon>ConfigSettings.cs</DependentUpon>
  </Compile>
</ItemGroup>
```

File: src/StackExchange.Exceptional.Shared/Settings.cs [hunk 2]
```diff
@@ -133,13 +133,13 @@ public class IgnoreSettings
...
             /// </summary>
-            public List<Regex> Regexes { get; set; }
+            public List<Regex> Regexes { get; set; } = new List<Regex>();
 
             /// <summary>
             /// Types collection for errors to ignore.  
             /// Any errors with a Type matching any name here will not be logged.
             /// </summary>
-            public HashSet<string> Types { get; set; }
+            public HashSet<string> Types { get; set; } = new HashSet<string>();
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.Shared/Settings.cs]: Perform a rebase on the other commit, so that these changes are folded out please? They're merged into master already.

File: src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs [hunk 3]
```diff
@@ -32,11 +34,10 @@ public static IApplicationBuilder UseExceptional(this IApplicationBuilder builde
         /// <param name="builder">The <see cref="IApplicationBuilder"/> instance this method extends.</param>
         /// <param name="applicationName">Application name for this error log.</param>
         /// <exception cref="ArgumentNullException"><paramref name="builder"/> or <paramref name="applicationName"/> is <c>null</c>.</exception>
-        public static IApplicationBuilder UseExceptional(this IApplicationBuilder builder, string applicationName)
+        public static IApplicationBuilder UseExceptional(this IApplicationBuilder builder, IConfiguration configuration, string applicationName)
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs]: Let's see if we can do this cleaner, e.g. can IConfiguration be injected or accessed another way? Passing it in seems very odd here, pretty sure there's a way around (I can take a stab this week)

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            /// <summary>
+            /// The Pattern to match on the exception message.
+            /// </summary>
+            public string Pattern { get; set; }
+
+            private Regex _patternRegEx;
+            /// <summary>
+            /// <see cref="Regex"/> object representing the pattern specified, compiled once for use against all future exceptions.
+            /// </summary>
+            public Regex PatternRegex => _patternRegEx ?? (_patternRegEx = new Regex(Pattern, RegexOptions.IgnoreCase | RegexOptions.Singleline));
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs]: No need to memoize, would move this into the loop.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace StackExchange.Exceptional
+{
+    /// <summary>
+    /// The Settings element for Exceptional's configuration.
+    /// This is the legacy web.config settings, that only serve as an adapter to populate <see cref="Settings"/>.
+    /// </summary>
+    public partial class ConfigSettings
+    {
+        const string CONFIGSECTION_KEY = "Exceptional";
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: public here

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        public static void ConfigureSettings(IConfiguration configuration, Settings settings)
+        {
+            var configSettings = new ConfigSettings();
+            configSettings._exceptionalConfiguration = configuration.GetSection(CONFIGSECTION_KEY);
+            if (configSettings._exceptionalConfiguration.Value != null)
+            {
+                configSettings._exceptionalConfiguration.Bind(configSettings);
+                configSettings.Initialize(settings);
+            }
+        }
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: This should be a constructor IMO, that takes just an `IConfiguration`, then we have a method called `Populate(Settings settings)` on it. It'd be cleaner and support more use cases, including caching, changing, and saving.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        /// </summary>
+        public string ApplicationName { get; set; }
+
+        /// <summary>
+        /// The regular expression pattern of data keys to include. 
+        /// For example, "Redis.*" would include all keys that start with Redis.
+        /// </summary>
+        public string DataIncludePattern { get; set; }
+
+        public bool ConfigSectionExists(string setting)
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Can remove `Config` prefix on these, we're all config friends here!

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            return _exceptionalConfiguration.GetChildren().Any(x => x.Key == setting);
+        }
+
+        public void Initialize(Settings settings)
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Would call this `Populate()`

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        {
+            // Main settings
+            settings.ApplicationName = ApplicationName;
+            if (DataIncludePattern.HasValue())
+            {
+                settings.DataIncludeRegex = new Regex(DataIncludePattern, RegexOptions.Singleline | RegexOptions.Compiled);
+            }
+
+            if (ConfigSectionExists(nameof(Email)))
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: I'd do this in the constructor, if it's `null`, just leave it `null`. Then these calls become `Email?.Populate(settings);` (change names there too please, and make them `internal`).

This way, if a user deserializes settings, changes one thing, and re-serializes, we don't spit out a bunch of defaults or empty collections as a result. Also: less code :)

File: src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs [hunk 10]
```diff
@@ -15,12 +16,13 @@ public static class ExceptionalBuilderExtensions
         /// <param name="builder">The <see cref="IApplicationBuilder"/> instance this method extends.</param>
         /// <param name="configureSettings">The action configuring Exceptional settings.</param>
         /// <exception cref="ArgumentNullException"><paramref name="builder"/> or <paramref name="configureSettings"/> is <c>null</c>.</exception>
-        public static IApplicationBuilder UseExceptional(this IApplicationBuilder builder, Action<Settings> configureSettings)
+        public static IApplicationBuilder UseExceptional(this IApplicationBuilder builder, IConfiguration configuration, Action<Settings> configureSettings)
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs]: Same as below, let's see if we can get rid of this argument. Don't focus on this, it's a usability optimization but we'll see what we can do. Right now, this makes it non-optional and that's a weird overload pattern...but I can refactor those.

File: src/StackExchange.Exceptional.AspNetCore/Extensions.cs [hunk 11]
```diff
@@ -179,17 +179,16 @@ NameValueCollection TryGetCollection(Func<HttpRequest, IEnumerable<KeyValuePair<
...
+                if (formFilters?.Count > 0)
                 {
-                    if (kv.Value != null)
+                    foreach (var kv in formFilters)
                     {
-                        error.Form[kv.Key] = kv.Value;
+                        if (kv.Value != null)
+                        {
+                            error.Form[kv.Key] = kv.Value;
+                        }
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/Extensions.cs]: Same: rebase so these go bye bye pretty please.

File: src/StackExchange.Exceptional.AspNetCore/StackExchange.Exceptional.AspNetCore.csproj (hunk 1 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/StackExchange.Exceptional.AspNetCore.csproj]: Sure! Thats awesome. Didn´t know that this is working already! Fixed in 8266281

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs (hunk 4 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs]: Sounds good to me. Changed it in both AspNetCore and Exceptional for classical mvc in ab3dcc3

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 7 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Fixed in 9325641

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 9 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Great idea! Really love this. Committed in 5e2f97a

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 8 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Changed in 088f265

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 6 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Sounds good to me. Changed in b0e02c8 Please take a look if this fits what you are talking about

File: src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs (hunk 3 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ExceptionalBuilderExtensions.cs]: This is not the best approach. Agreed. One of the possible solutions that I can think of is instantiating `ConfigSettings` in `ctor` of `ExceptionalMiddleWare`. 

So basically we would add this call in the Startup class if the user likes to use the Configuration for exceptional:

```csharp
services.Configure<ConfigSettings>(Configuration.GetSection("Exceptional")
```

We could then inject `ConfigSections` in `ExceptionMiddleware` ctor and call `Populate(settings)` in there.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 5 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Moved the initialization of the Configuration to bind outside our lib in f0aa05a

**Comment by StefanKert** [comment]
Thanks for the feedback Nick! Really appreciate it. Fixed most of the issues. I need to do the rebase, but therefore i need a little time to get my git skills straight ;)

There is just the ApplicationBuilder exception thing that we need to perform some steps on. I added a proposal in the comments section, but just to be safe:

This is not the best approach. Agreed. One of the possible solutions that I can think of is instantiating `ConfigSettings` in `ctor` of `ExceptionalMiddleWare`. 

So basically we would add this call in the Startup class if the user likes to use the Configuration for exceptional:

```csharp
services.Configure<ConfigSettings>(Configuration.GetSection("Exceptional")
```

We could then inject `ConfigSections` in `ExceptionMiddleware` ctor and call `Populate(settings)` in there.

**Comment by StefanKert** [comment]
Another possible solution would be to extend the `IServiceCollection`:

```csharp
public static IServiceCollection ConfigureExceptionalSettings(this IServiceCollection services, IConfiguration config)
{
 //nullchecks
  var configSettings = new ConfigSettings(configuration);
  configSettings.Populate(Settings.Current);
}
```

This way we would instantiate the current settings with the give configuration. The user has to add a call to this in his `ConfigureServices` Method similar to this:

```csharp
services.ConfigureExceptionalSettings(Configuration.GetSection("Exceptional");
```

We could also add it to the DI Container so that we can retrieve it by adding `IOptions<ConfigSettings>` to any ctor.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            /// <summary>
+            /// Regular expressions collection for errors to ignore.  
+            /// Any errors with a .ToString() matching any regular expression here will not be logged.
+            /// </summary>
+            public List<IgnoreRegex> Regexes { get; set; } = new List<IgnoreRegex>();
+
+            /// <summary>
+            /// Types collection for errors to ignore.  Any errors with a Type matching any name here will not be logged
+            /// </summary>
+            public List<IgnoreType> Types { get; set; } = new List<IgnoreType>();
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.IgnoreErrors.cs]: Since we're not looking for compatibility with existing `web.config`s here (the layout for ASP.NET non-Core), let's simplify this. These can simply be `List<string>` for both properties, no defaults - just `null`.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+            /// <summary>
+            /// The size of this error log, either how many to keep or how many to display depending on type.
+            /// </summary>
+            public int Size { get; set; } = 200;
+
+            /// <summary>
+            /// The duration in seconds of error groups to roll-up, similar errors within this timespan will be shown as duplicates.
+            /// </summary>
+            public int RollupSeconds { get; set; } = 600;
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs]: Let's make all of these nullable (`int?`), since they're just overrides to defaults here...we don't want to maintain defaults here as well.

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+            /// <summary>
+            /// The connection string to use on database based error stores.  If provided, ConnectionStringName is ignored.
+            /// </summary>
+            public string ConnectionString { get; set; }
+
+            /// <summary>
+            /// The name of the connection string to use from the application's configuration.
+            /// </summary>
+            public string ConnectionStringName { get; set; }
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs]: `#if !NETSTANDARD2_0` wrap here please (I'll change all these later if we add an earlier `netstandard` build)

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.Email.cs [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+            /// <summary>
+            /// The SMTP password to use, if authentication is needed.
+            /// </summary>
+            public string SMTPPassword { get; set; }
+
+            /// <summary>
+            /// Whether to use SSL when sending via SMTP.
+            /// </summary>
+            public bool SMTPEnableSSL { get; set; }
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.Email.cs]: Nullables on all - only set if not null in `Populate()`.

File: src/StackExchange.Exceptional.AspNetCore/ExceptionalMiddleware.cs [hunk 16]
```diff
@@ -49,7 +49,7 @@ public class ExceptionalMiddleware
         /// <summary>
         /// Executes the Exceptional-wrapped middleware.
         /// </summary>
-        public async Task Invoke(HttpContext context)
+        public async Task Invoke(HttpContext context)
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ExceptionalMiddleware.cs]: Minor noise here just needs a revert

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+            settings.ApplicationName = ApplicationName ?? settings.ApplicationName;
+            if (DataIncludePattern.HasValue())
+            {
+                settings.DataIncludeRegex = new Regex(DataIncludePattern, RegexOptions.Singleline | RegexOptions.Compiled);
+            }
+
+            Email?.Populate(settings);
+            ErrorStore?.Populate(settings);
+            IgnoreErrors?.Populate(settings);
+            LogFilters.Populate(settings);
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Missed a `?.` here :)

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+                }
+            }
+        }
+
+        public LogFilterSettings LogFilters { get; set; }
+        public class LogFilterSettings
+        {
+            public List<LogFilter> Form { get; set; } = new List<LogFilter>();
+
+            public List<LogFilter> Cookies { get; set; } = new List<LogFilter>();
```
Comment (NickCraver) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: These can just be `Dictionary<string, string>`, the reason the non-Core version isn't is *purely* because of backwards compatibility with existing configs...we can be cleaner here and eliminate the `LogFilter` class :)

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.Email.cs (hunk 15 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.Email.cs]: Done in 58e2ead

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs (hunk 13 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs]: Done in 194970e

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs (hunk 14 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.ErrorStore.cs]: Sorry :) missed that one. Added in 	99dce6c

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 17 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: Haha :) good catch. Already fixed a few commits ago :)

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.LogFilters.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+{
+    internal partial class ConfigSettings
+    {
+        public LogFilterSettings LogFilters { get; set; } 
+
+        public class LogFilterSettings
+        {
+            public List<LogFilter> Form { get; set; } = new List<LogFilter>();
+
+            public List<LogFilter> Cookies { get; set; } = new List<LogFilter>();
```
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.LogFilters.cs]: Awesome :) had this in when I was starting with the Settings impl, but removed it again because I wanted to be as close to the original as possible :) back in in b1a1513

File: src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs (hunk 18 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ConfigSettings.cs]: We´re getting closer to a really small class 👍 b1a1513

**Comment by StefanKert** [comment]
I´ve already created an example file that I used for testing, need to get some comments in it! 

Just another thing regarding IgnoreSettings and probably also LogFilters. If We call `Populate` again any time during the app/webapp is running, we would add up all Filters and IgnoreSettings to the current list in `Settings`. Shouldn´t we add a check if these items already exist in the list? I don´t want to clear the list at all because if somebody has already added some items they get also cleared out so I think here it would be better to check if these items already exist. Maybe you got different thoughts on this? Please let me know, but I think this could be a problem if we don´t check what to do now.

**Comment by NickCraver** [comment]
@StefanKert That's why the underlying implementations are de-duped collection types ;) IgnoreSettings.Regexes as the exception to this (just changed), so shouldn't be an issue anymore :)

**Comment by StefanKert** [comment]
@NickCraver Awesome :) try to take a look if we could use the whole `ConfigSettings` class as internal. What are you thinking about providing a separate `Configure` Method?

```csharp
public static IServiceCollection ConfigureExceptionalSettings(this IServiceCollection services, IConfiguration config)
{
 //nullchecks
  var configSettings = new ConfigSettings(configuration);
  configSettings.Populate(Settings.Current);
  // add to di container
 return services;
}
```

Could be the replacement for the `IConfiguration` parameter.

**Comment by NickCraver** [comment]
I think what you have is better, it's not about being clever here, but about user expectations. ASP.NET Core has certain patterns, the `.GetSection()` approach is the normal pattern...so we go with the flow. I'll just add some overloads when we're done without the param, no big deal.

**Comment by StefanKert** [comment]
Ok. I can totally add up to this, but I thought about using similar approach to the recommended way as it is in aspnet core. For example if you´d like to load your configuration into a setings class you would implement it like this:
```csharp
    public void ConfigureServices(IServiceCollection services)
    {
        // Adds services required for using options.
        services.AddOptions();

        // Register the IConfiguration instance which MyOptions binds against.
        services.Configure<MySubOptions>(Configuration.GetSection("subsection"));

        // Add framework services.
        services.AddMvc();
    }
```

So we got the `Configure` Method here that populates the `MyOptions` class and adds it to the DI Container. So the call for our method would probably look like this:

```csharp
    public void ConfigureServices(IServiceCollection services)
    {
        // Adds services required for using options.
        services.AddOptions();

        // Register the IConfiguration instance which MyOptions binds against.
        services.ConfigureExceptionalSettings(Configuration.GetSection("Exceptional"));

        // Add framework services.
        services.AddMvc();
    }
```

Just wanted to throw this in so you get what I mean with that :) I am totally fine with the overload approach, just wanted to leave this in here :)

**Comment by StefanKert** [comment]
@NickCraver I tested if we could make the `ConfigSettings` class internal only, but right now this is not working. aspnet/Configuration#394 and if I understand the comments correctly this will come in 2.1? There is also another package that lets you set all settings even if they have no setter, but this would introduce another dependency on a package that is out of our control and even not in control of Microsoft.  So I suggest that we use public for all the properties that need them and wait for the new Configuration release.

File: src/StackExchange.Exceptional.AspNetCore/ExceptionalMiddleware.cs (hunk 16 above)
Comment (StefanKert) [inline src/StackExchange.Exceptional.AspNetCore/ExceptionalMiddleware.cs]: Removed white spaces. Don´t know if I should remove the commit at all or only whitespaces. Please let me know :)

**Comment by NickCraver** [comment]
@StefanKert whelp, that sucks. I appreciate the linked issue, I'll help track that down and see if we can get the change in. I'm ton on whether to use a custom binder, or waste even more time and duplication in restoring the comments...I'm leaning towards the binder, but I can take that separately. Let me look around for options.

**Comment by NickCraver** [comment]
@StefanKert the `internal` class seems to be working here, I'll test out more tweaks but from my reading of the default binder only non-public *properties* (or constructors) would have any issues. I think we're all good.

Merged into master and moved to `.Internal` to cleanup, same with extension changes, thank you!!

**Comment by StefanKert** [comment]
@NickCraver I guess I should have tested a little bit more :) only tested all internal. Awesome that this is working now.

Great to hear :) if you need any further help just ping me here or on Twitter.
